internal/services/authorization/role_definition_resource.go

```diff
@@ -10,348 +10,411 @@ import (
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Authorization.RoleDefinitionsClient
 
-	permissionsRaw := d.Get("permissions").([]interface{})
-	permissions := expandRoleDefinitionPermissions(permissionsRaw)
-	assignableScopes := expandRoleDefinitionAssignableScopes(d)
+			var config RoleDefinitionModel
+			if err := metadata.Decode(&config); err != nil {
+				return fmt.Errorf("decoding %+v", err)
+			}
 
-	if d.IsNewResource() {
-		existing, err := client.Get(ctx, scope, roleDefinitionId)
-		if err != nil {
+			roleDefinitionId := config.RoleDefinitionId
+			if roleDefinitionId == "" {
+				uuid, err := uuid.GenerateUUID()
+				if err != nil {
+					return fmt.Errorf("generating UUID for Role Assignment: %+v", err)
+				}
+				roleDefinitionId = uuid
+			}
+
+			existing, err := client.Get(ctx, config.Scope, roleDefinitionId)
+			if err != nil && !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for presence of existing Role Definition ID for %q (Scope %q)", config.Name, config.Scope)
+			}
 			if !utils.ResponseWasNotFound(existing.Response) {
-				return fmt.Errorf("checking for presence of existing Role Definition ID for %q (Scope %q)", name, scope)
+				importID := parse.RoleDefinitionID{
+					RoleID: roleDefinitionId,
+					Scope:  config.Scope,
+				}
+				return metadata.ResourceRequiresImport(r.ResourceType(), importID)
 			}
-		}
 
-		if existing.ID != nil && *existing.ID != "" {
-			importID := fmt.Sprintf("%s|%s", *existing.ID, scope)
-			return tf.ImportAsExistsError("azurerm_role_definition", importID)
-		}
-	}
+			properties := authorization.RoleDefinition{
+				RoleDefinitionProperties: &authorization.RoleDefinitionProperties{
+					RoleName:         &config.Name,
+					Description:      &config.Description,
+					RoleType:         pointer.To("CustomRole"),
+					Permissions:      pointer.To(expandRoleDefinitionPermissions(config.Permissions)),
+					AssignableScopes: pointer.To(expandRoleDefinitionAssignableScopes(config)),
+				},
+			}
 
-	properties := authorization.RoleDefinition{
-		RoleDefinitionProperties: &authorization.RoleDefinitionProperties{
-			RoleName:         utils.String(name),
-			Description:      utils.String(description),
-			RoleType:         utils.String(roleType),
-			Permissions:      &permissions,
-			AssignableScopes: &assignableScopes,
-		},
-	}
+			if _, err := client.CreateOrUpdate(ctx, config.Scope, roleDefinitionId, properties); err != nil {
+				return err
+			}
 
-	if _, err := client.CreateOrUpdate(ctx, scope, roleDefinitionId, properties); err != nil {
-		return err
+			read, err := client.Get(ctx, config.Scope, roleDefinitionId)
+			if err != nil {
+				return err
+			}
+			if read.ID == nil || *read.ID == "" {
+				return fmt.Errorf("Cannot read Role Definition ID for %q (Scope %q)", config.Name, config.Scope)
```

Comment:
Beginning of error messages should be lowercased
```suggestion
				return fmt.Errorf("cannot read Role Definition ID for %q (Scope %q)", config.Name, config.Scope)
```